cpp/src/gandiva/jni/jni_common.cc

```diff
@@ -572,6 +586,192 @@ void releaseProjectorInput(jbyteArray schema_arr, jbyte* schema_bytes,
+          types::Schema schema;
+          types::ExpressionList exprs;
+          ExpressionVector expr_vector;
+          SchemaPtr schema_ptr;
+          FieldVector ret_types;
+          gandiva::Status status;
+          auto mode = gandiva::SelectionVector::MODE_NONE;
+
+          ConfigurationBuilder configuration_builder;
+          std::shared_ptr<Configuration> config = configuration_builder.build();
+
+          if (!ParseProtobuf(schema_bytes, schema_len, &schema)) {
+            std::cerr << "[Gandiva Cache Prewarm] Failed to parse protobuf for schema"
+                      << std::endl;
+            continue;
+          }
+
+          if (!ParseProtobuf(exprs_bytes, exprs_len, &exprs)) {
+            std::cerr << "[Gandiva Cache Prewarm] Failed to parse protobuf for expr list"
+                      << std::endl;
+            continue;
+          }
+
+          // convert types::Schema to arrow::Schema
+          schema_ptr = ProtoTypeToSchema(schema);
+          if (schema_ptr == nullptr) {
+            std::cerr << "[Gandiva Cache Prewarm] Failed to convert protobuf schema to "
+                         "arrow type"
+                      << std::endl;
+            continue;
+          }
+
+          // create Expression out of the list of exprs
+          for (int i = 0; i < exprs.exprs_size(); i++) {
+            ExpressionPtr root = ProtoTypeToExpression(exprs.exprs(i));
+
+            if (root == nullptr) {
+              std::cerr << "[Gandiva Cache Prewarm] Failed to convert protobuf expr to "
+                           "arrow type"
+                        << std::endl;
+              continue;
+            }
+
+            expr_vector.push_back(root);
+            ret_types.push_back(root->result());
+          }
+
+          status = Projector::Make(schema_ptr, expr_vector, mode, config, &projector);
+
+          if (!status.ok()) {
+            std::cerr << "[Gandiva Cache Prewarm] Failed to create a projector module";
+            continue;
+          }
+          std::cout << "[Gandiva Cache Prewarm] Built and cached a projector from the "
+                       "expression and schema in the file"
+                    << std::endl;
+        }
+      } catch (const std::exception& ex) {
+        std::cerr << "[Gandiva Cache Prewarm] " << dir_iter->path().filename() << " "
+                  << ex.what() << std::endl;
+      }
+    }
+
+  } catch (const std::exception& ex) {
+    std::cerr << "[Gandiva Cache Prewarm] Failed prewarming the cache: " << ex.what()
+              << std::endl;
+  }
+}
+
+void CacheExpressionAndSchemaForPrewarmOnStartup(char* schema_bytes, int32_t schema_len,
+                                                 char* exprs_bytes, int32_t exprs_len) {
+  try {
+    auto env_path = std::getenv(kEnvPrewarmCacheDir);
+    if (env_path == nullptr) {
+      return;
+    }
+
+    fs::path path(env_path);
+    if (!fs::is_directory(path) && !fs::create_directories(path)) {
```

Comment:
is this thread safe? two threads trying to create the directory at the same time